dilated-file-transfer.md

```diff
@@ -0,0 +1,499 @@
+
+Rejected idea: Version message, because we already do version negotiation via mailbox features.
+
+Rejected idea: Offer/Answer messages via the control channel: we need to open a subchannel anyway and the subchannel-IDs are not intended to be part of the Dilation public API.
+
+
+### Control Channel Messages
+
+Each side MAY send a free-form text message at any time.
+These messages look like:
+
+```python
+class Message:
+    message: str     # unicode string
+    kind: str = "text"
+```
+
+
+### Making an Offer
+
+Either side MAY propose any number of Offers at any time after the connection is set up.
+If the other peer specified `"mode": "send"` then this peer MUST NOT make any Offers.
+If this peer specified `"mode": "receive"` then this peer MUST NOT make any Offers.
+
+To make an Offer the peer opens a subchannel.
+Recall from the Dilation specification that subchannels are _record_ pipes (not simple byte-streams).
+
+All records on the subchannel begin with a single byte indicating the kind of message.
+Any additional bytes are a kind-dependent payload.
+
+The following kinds of messages exist (as indicated by the first byte):
+* 1: msgpack-encoded `FileOffer` message
+* 2: msgpack-encoded `DirectoryOffer` message
+* 3: msgpack-encoded `OfferAccept` message
+* 4: msgpack-encoded `OfferReject` message
+* 5: raw file data bytes
+
+All other byte values are reserved for future use and MUST NOT be used.
+
+    XXX: maybe spec [0, 128) as reserved, and [128, 255) for "experiments"?
+
+The sender that opened the new subchannel MUST immediately send one of the two kinds of offer messages.
+
+To offer a single file (with message kind `1`):
+
+```python
+class FileOffer:
+    filename: str   # filename (no path segments)
+    timestamp: int  # Unix timestamp (seconds since the epoch in GMT)
+    bytes: int      # total number of bytes in the file
+```
+
+To offer a directory tree of many files (with message kind `2`):
+
+```python
+class DirectoryOffer:
+    base: str              # unicode pathname of the root directory (i.e. what the user selected)
+    size: int              # total number of bytes in _all_ files
+    files: list[list[str]] # a list containing relative paths for each file
+                           # each relative path is a sequence of unicode strings (relative to "base")
+```
+
+The filenames in the `"files"` list are sequences of unicode path-names and are relative to the `"base"` from the `DirectoryOffer` (but NOT including that part).
+Note that a `FileOffer` message also precedes each file in the Directory when the data is streamed.
+The files MUST be streamed in the same order they appear in the `files` list.
+The last segment of each entry in the filename list MUST match the `"filename"` of the `FileOffer` message.
+
+For example:
+
+```python
+DirectoryOffer(
+    base="project",
+    size: 165,
+    files=[
+        ["README"],
+        ["src", "hello.py"],
+    ]
+)
+```
```

Comment:
This opens a few questions for me:

- What is the size of a directory? The sum of all files? Does it include metadata size?
- Is it really required to send a list of all files beforehand? This might actually be more difficult than expected due to multiple reasons:
  - What if during the transfer the list of files in the directory changes? Files might get added/removed.
  - What if the size of a single file changes before opening the file for sending?
  - Recursing the directory before the transfer can be slow
- The only way to prevent this race condition is to open *all* files at once before sending the directory offer and not letting go of the file handles until the files have finished sending. The open file limit would then limit the number of files that can be sent at once. As this limit might vary a lot and depends on other applications on the system this really shouldn't grow linearly with the number of files to be sent.

Reply:
Yes, I intended "size" to be "sum of all files". I'm not sure what you mean by "metadata" but it would just be the on-disk sizes.

The idea of sending the list beforehand is so the (presumably human) on the other end has some idea what they're agreeing to receive. That is, the question can be a bit better than `Receive "Documents/" y/n?`

_What if the directory contents (list of files) changes_: either fail the transfer or ignore the new ones? You'd have to fail if a promised file "went away" according to the above.

_What if the size changes_: we could spec the "up-front"  / directory size as an estimate, so that only the "header" size matters (which should be late enough that you're actually opening the file)?

Slow to recurse: really only if there's a _lot_ of files. What else would you suggest? How can a meaningful but "open-ended" offer happen?

Reply:
Sending beforehand which files are going to be sent is more or less required, if only because the receiver wants to know what they can expect before acknowledging it. File system smear is a big issue though, so even though we can't reasonably prevent it we must provide mechanisms for detecting it (e.g. through integrity checks).

Reply:
This could be another reason to keep the `"permission": "yes"` mode or similar, because then the sender _doesn't_ have to do all that work and can just send files (if they know the receiver is saying "yes to all" up front).

Reply:
The best we can hope for is probably `stat()` i.e. size + mtime basically .. if that's what you mean by "integrity checks"?

Reply:
> _What if the size changes_: we could spec the "up-front" / directory size as an estimate, so that only the "header" size matters (which should be late enough that you're actually opening the file)?

That sounds reasonable

Reply:
How will checksumming be used to detect "filesystem smear"?

Reply:
Nevermind, it doesn't. At least not without having to read the entire folder from disk twice, which is a lot more expensive than just checking metadata. Then we can only do "best effort" by checking file size and mtime. (My confusion came from my proposal where for resumption, checksums can be used to detect file system smear across transfers.)

Still, checksums are something we likely want, both on the transferred bytes level as well as for the individual files, especially if we want to go down the path where all files are concatenated together (in order) into one byte stream.

Reply:
> I'm in favor of being precise about what is going to be sent up-front and then committing to that. Traversing a directory can indeed be slow sometimes*, but it will still be an order of magnitude faster than the current solution where the entire directory has to be zipped up-front.

If this is the conclusion here the specification should probably elaborate on those potential issues and offer guidance on how to proceed in all of those cases. For example:
- Ignore any files that are added after the fact
- Abort the transfer if any of the files have changed their size / mtime in between or are missing althogether

Reply:
added more words to that effect

Reply:
i also added a `FileAcknowledge` message with a file hash.

Reply:
I think this covers all cases now, thanks.